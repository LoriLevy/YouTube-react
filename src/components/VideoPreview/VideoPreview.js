/* file: src/components/VideoPreview/VideoPreview.js */
import React from "react";
import { Image } from "semantic-ui-react";
import "./VideoPreview.scss";
import {
  getVideoDurationString,
  formatNumberWithCommas
} from "../../utilities/format-number";
import moment from "moment";

function VideoPreview(props) {
  const {video} = props;
  const horizontal = props.horizontal ? "horizontal" : null;
  const duration = video.contentDetails ? video.contentDetails.duration : null;
  const videoDuration = getVideoDurationString(duration);
  const views = video.statistics.viewCount ? video.statistics.viewCount : "0";
  let viewCount = formatNumberWithCommas(views);
  // round the number of days since typically the returned value contains a decimal
  let daysAgo = Math.round(
    moment
      .duration(moment(new Date()).diff(moment(video.snippet.publishedAt)))
      .asDays()
  );

  // if it was one day ago, it should be "1 day ago"; if more than one day, it should be "days ago"
  const numberOfDays = daysAgo.toString();
  const singular = " day ago";
  const plural = " days ago ";
  let howMany = daysAgo === 1 ? numberOfDays.concat(singular) : numberOfDays.concat(plural);

  const thumbnaillUrl = `${video.snippet.thumbnails.medium.url}`

  return (
    <div className={["video-preview", horizontal].join(" ")}>
      <div className="image-container">
        <Image src={thumbnaillUrl} />
        <div className="time-label">
          <span>{videoDuration}</span>
        </div>
      </div>

      <div className="video-info">
        <div className="semi-bold show-max-two-lines">{video.snippet.localized.title}</div>
        <div className="video-preview-metadata-container">
          <div className="channel-title">{video.snippet.channelTitle}</div>
          <div>
            <span>{viewCount} views â€¢ {howMany}</span>
          </div>
        </div>
      </div>
    </div>
  );
}
export default VideoPreview;
