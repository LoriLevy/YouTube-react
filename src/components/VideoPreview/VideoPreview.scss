@import "../../styles/shared.scss";

.video-preview {
  display: grid; /* Grid parent */
  grid: 118px auto / 210px; /* Row height = 118px / Column width = 210px  */
  padding: 1rem 1.2rem;
}

.video-info {
  grid-row: 2 / 3; /* position: between second and third row */
  grid-column: 1 / 2; /* position: between first and second column */
  color: $text-color-dark;

  .video-preview-metadata-container {
    padding-top: 5px;
    font-size: 1.1rem;
  }
  .channel-title {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: 1.1rem;
  }
}

.show-max-two-lines {
  overflow: hidden;
  line-height: 1.4em;
  max-height: 2.8em;

  .semi-bold {
    font-weight: 600;
  }
}

.image-container {
  position: relative;
  grid-row: 1 / 2; /* position: between first and second row */
  grid-column: 1 / 2; /* position: between first and second column */

  /* Video duration label at bottom right */
  .time-label {
    position: absolute;
    background: $text-color-dark;
    bottom: 0;
    right: 0;
    opacity: 0.8;
    border-radius: 2px;
    font-weight: 500;
    color: $white;
    margin: 4px;
    padding: 2px 4px;
    line-height: 12px;
  }
}
