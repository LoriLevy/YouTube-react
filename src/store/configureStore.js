import {applyMiddleware, createStore, compose} from 'redux';
import reducer from './reducers';
import createSagaMiddleware from 'redux-saga';
import rootSaga from './sagas';

export function configureStore() {
    const sagaMiddleware = createSagaMiddleware();
    //added composeEnhancers for seeing history of actions
    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    const store = createStore(reducer, 
        composeEnhancers(applyMiddleware(sagaMiddleware))
    );

    sagaMiddleware.run(rootSaga);
    return store;
}

/* returns __REDUX_DEVTOOLS_EXTENSION_COMPOSE__ */