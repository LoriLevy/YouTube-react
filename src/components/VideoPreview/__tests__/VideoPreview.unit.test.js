import React from 'react';
import { shallow } from 'enzyme';
import VideoPreview from '../VideoPreview';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {  Provider  } from 'react-redux';

const middleware = [thunk];
const mockStore = configureStore(middleware);
const initialState = {
  example: {}
};
const store = mockStore(initialState);

describe(VideoPreview, () => {
  test('it renders vertically', () => {
    const wrapper = shallow(<Provider store={store}><VideoPreview/></Provider>);
    expect(wrapper).toMatchSnapshot();
  });
  test('it renders horizontally', () => {
    const wrapper = shallow(<Provider store={store}><VideoPreview horizontal={true}/></Provider>);
    expect(wrapper).toMatchSnapshot();
  });
});
