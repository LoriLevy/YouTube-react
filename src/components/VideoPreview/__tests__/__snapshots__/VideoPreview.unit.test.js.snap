// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPreview it renders horizontally 1`] = `
<div
  className="video-preview horizontal"
>
  <div
    className="image-container"
  >
    <Image
      as="img"
      src="http://via.placeholder.com/210x118"
      ui={true}
    />
    <div
      className="time-label"
    >
      <span>
        05:22
      </span>
    </div>
  </div>
  <div
    className="video-info"
  >
    <div
      className="semi-bold show-max-two-lines"
    >
      Video title
    </div>
    <div
      className="video-preview-metadata-container"
    >
      <div
        className="channel-title"
      >
        Channel title
      </div>
      <div>
        <span>
          1699 views • 12 days ago
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`VideoPreview it renders vertically 1`] = `
<div
  className="video-preview "
>
  <div
    className="image-container"
  >
    <Image
      as="img"
      src="http://via.placeholder.com/210x118"
      ui={true}
    />
    <div
      className="time-label"
    >
      <span>
        05:22
      </span>
    </div>
  </div>
  <div
    className="video-info"
  >
    <div
      className="semi-bold show-max-two-lines"
    >
      Video title
    </div>
    <div
      className="video-preview-metadata-container"
    >
      <div
        className="channel-title"
      >
        Channel title
      </div>
      <div>
        <span>
          1699 views • 12 days ago
        </span>
      </div>
    </div>
  </div>
</div>
`;
